jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/AbstractHTTP2ServerConnectionFactory.java

```diff
@@ -182,14 +181,43 @@ public void setMaxSettingsKeys(int maxSettingsKeys)
         this.maxSettingsKeys = maxSettingsKeys;
     }
 
+    /**
+     * @return null
+     * @deprecated use {@link #getRateControlFactory()} instead
+     */
+    @Deprecated
     public RateControl getRateControl()
     {
-        return rateControl;
+        return null;
     }
 
+    /**
+     * @param rateControl ignored
+     * @throws UnsupportedOperationException when invoked
+     * @deprecated use {@link #setRateControlFactory(RateControl.Factory)} instead
+     */
+    @Deprecated
     public void setRateControl(RateControl rateControl)
     {
-        this.rateControl = rateControl;
+        throw new UnsupportedOperationException();
```

Comment:
We have users who have been told to directly call this.  It should add a factory that wraps the passed rateControl.

Reply:
That won't work, as the same `RateControl` object will be used for all connections, which is what we want to fix.

Reply:
I said add a factory not an instance!

Reply:
ie:
```java
this.setRateControlFactory(new WindowRateControl.Factory(rateControl));
```

Reply:
@gregw I understood what you meant, but still does not work. That factory you are passing it will be implemented like this:

```java
class WindowRateControl.Factory ... {
    private final RateControl _rateControl;

    WindowRateControl.Factory(RateControl rateControl) {
        _rateControl = rateControl;
    }

    RateControl newRateControl(EndPoint endPoint) {
        return _rateControl;
    }
}
```

For every different `EndPoint` we are returning the same object (the one we passed to the constructor), which is not what we want.
There is no way to create a new `RateControl` from a prototype instance (e.g. using `clone()` or something).

Reply:
@gregw ok, I have tried to minimize the impact for `setRateControl()` so now if it receives exactly a `WindowRateControl`, it extracts the rate and sets a factory.
If it's not precisely a `WindowRateControl`, throws `UnsupportedOperationException`.